homeassistant/components/tesla/__init__.py

```diff
@@ -169,72 +181,110 @@ async def async_setup_entry(hass, config_entry):
     username = config_entry.title
-    hass.data[DOMAIN].pop(config_entry.entry_id)
-    _LOGGER.debug("Unloaded entry for %s", username)
-    return True
+    if unload_ok:
+        hass.data[DOMAIN].pop(config_entry.entry_id)
+        _LOGGER.debug("Unloaded entry for %s", username)
+        return True
+    return False
 
 
 async def update_listener(hass, config_entry):
     """Update when config_entry options update."""
-    controller = hass.data[DOMAIN][config_entry.entry_id]["controller"]
+    controller = hass.data[DOMAIN][config_entry.entry_id]["coordinator"].controller
     old_update_interval = controller.update_interval
     controller.update_interval = config_entry.options.get(CONF_SCAN_INTERVAL)
-    _LOGGER.debug(
-        "Changing scan_interval from %s to %s",
-        old_update_interval,
-        controller.update_interval,
-    )
+    if old_update_interval != controller.update_interval:
+        _LOGGER.debug(
+            "Changing scan_interval from %s to %s",
+            old_update_interval,
+            controller.update_interval,
+        )
+
+
+class TeslaDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching Tesla data."""
+
+    def __init__(self, hass, *, config_entry, controller):
+        """Initialize global Tesla data updater."""
+        self.controller = controller
+        self.config_entry = config_entry
+
+        update_interval = timedelta(seconds=MIN_SCAN_INTERVAL)
+
+        super().__init__(
+            hass, _LOGGER, name=DOMAIN, update_interval=update_interval,
+        )
+
+    async def _async_update_data(self):
+        """Fetch data from API endpoint."""
+        if self.controller.is_token_refreshed():
+            (refresh_token, access_token) = self.controller.get_tokens()
+            _async_save_tokens(
+                self.hass, self.config_entry, access_token, refresh_token
+            )
+            _LOGGER.debug("Saving new tokens in config_entry")
+
+        try:
+            # Note: asyncio.TimeoutError and aiohttp.ClientError are already
+            # handled by the data update coordinator.
+            async with async_timeout.timeout(30):
+                return await self.controller.update()
+        except TeslaException as err:
+            raise UpdateFailed(f"Error communicating with API: {err}")
 
 
 class TeslaDevice(Entity):
     """Representation of a Tesla device."""
 
-    def __init__(self, tesla_device, controller, config_entry):
+    def __init__(self, tesla_device, coordinator):
         """Initialise the Tesla device."""
         self.tesla_device = tesla_device
-        self.controller = controller
-        self.config_entry = config_entry
-        self._name = self.tesla_device.name
-        self.tesla_id = slugify(self.tesla_device.uniq_name)
-        self._attributes = {}
-        self._icon = ICONS.get(self.tesla_device.type)
+        self.coordinator = coordinator
+        self._attributes = self.tesla_device.attrs.copy()
```

Comment:
We want to see all attributes set. Please set them explicitly instead.

Reply:
For those that are static that is probably fine. However, for those that are dynamic that introduces additional complexity since the underlying library already creates all the HA properties for HA use. That means a single update requires updating both the underlying library and in HA which adds administrative burdens.

Reply:
It's a requirement.

Reply:
If that's required, then I'm going to move the complexity out of telsajsonpy into HA then to avoid the double updates. It's stupid that the library has a whole class of HA objects.

Reply:
That will be a large PR that will come later after these minor edits.

Reply:
I'm not sure about moving complexity into home assistant but I'll be happy to review the suggestion.

Reply:
I'm not planning to maintain that level of complexity in teslajsonpy anymore. I've tried for long enough and frankly it's either it gets fixed in HA or in teslajsonpy or I'm not going to bother anymore.